File: 프로그래머스/3/81303. 표 편집/표 편집.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using ai3 = array<int, 3>;

string solution(int n, int k, vector<string> cmd) {
    vector<int> pre(n, -1), nxt(n, -1);
    vector<char> ans(n, 'O');
    stack<ai3> st;
    for(int i = 0; i < n; i++){
        pre[i] = i-1;
        nxt[i] = i+1;
    }
    nxt[n-1] = -1;
    for(auto& c : cmd){
        if(c[0] == 'D'){
            int v = stoi(c.substr(2));
            while(v--) k = nxt[k];
        }
        else if(c[0] == 'U'){
            int v = stoi(c.substr(2));
            while(v--) k = pre[k];
        }
        else if(c[0] == 'C'){
            st.push({pre[k], k, nxt[k]});
            if(pre[k] != -1) nxt[pre[k]] = nxt[k];
            if(nxt[k] != -1) pre[nxt[k]] = pre[k];
            if(nxt[k] == -1) k = pre[k];
            else k = nxt[k];
        }
        else{
            auto [pi, cur, ni] = st.top(); st.pop();
            if(pi != -1) nxt[pi] = cur;
            if(ni != -1) pre[ni] = cur;
        }
    }
    while(st.size()){
        auto [pi, cur, ni] = st.top(); st.pop();
        ans[cur] = 'X';
    }
    return string(ans.begin(), ans.end());
}

File: 프로그래머스/3/81303. 표 편집/표 편집.py
def solution(n, k, cmd):
    ans, pre, nxt, st = ['O']*n, [-1]*n, [-1]*n, []
    for i in range(n): pre[i], nxt[i] = i-1, i+1
    nxt[-1] = -1
    for c in cmd:    
        if c[0] == 'D': 
            v = int(c.split()[1])
            for _ in range(v): k = nxt[k]
        elif c[0] == 'U':
            v = int(c.split()[1])
            for _ in range(v): k = pre[k]
        elif c[0] == 'C':
            st.append((pre[k], k, nxt[k]))
            if pre[k] != -1: nxt[pre[k]] = nxt[k]
            if nxt[k] != -1: pre[nxt[k]] = pre[k]
            if nxt[k] == -1: k = pre[k]
            else: k = nxt[k]
        else:
            pi, cur, ni = st.pop()
            if pi != -1: nxt[pi] = cur
            if ni != -1: pre[ni] = cur
    for _, cur, _ in st: ans[cur] = 'X'
    return ''.join(ans)